### ISSUE TITLE
Carefully consider behavior of nullish range options

### ISSUE THREAD

**Issue by vweevers** [Level/leveldown#515]
Though the logic in f26764762566948e0707c85068c7b0de9b50add3 felt "right", I'm afraid it could break the ecosystem. It could be incompatible with popular tools like [`ltgt`](https://www.npmjs.com/package/ltgt). For example, `ltgt.lowerBound({})` returns `undefined`.

If you're wondering why this doesn't cause problems for `encoding-down`, it has special logic for nullish. It doesn't touch them in the serialization step or [in the encoding step](https://github.com/Level/codec/blob/2f7e033b6c5b04d57b49030e2bf74f22ebc4e58b/lib/encodings.js#L63) (of builtin encodings). This avoids `{ gt: undefined }` turning into `{ gt: 'undefined' }`.

I'm leaning towards reverting f26764762566948e0707c85068c7b0de9b50add3, perhaps even the `abstract-leveldown` behavior (so that nullish means "not defined" again everywhere), but this needs some tests and thinking. If possible, I want to avoid needing special logic for nullish in every `abstract-leveldown` implementation. Because really, the only odd duck is `encoding-down`.

![giphy 9](https://user-images.githubusercontent.com/3055345/47260683-84b40600-d4c0-11e8-8822-ca23b476574c.gif)

Perhaps, if we revert the `abstract-leveldown` behavior, a new abstract `_isValidTarget()` method (bikeshed name) could allow `encoding-down` to enable nullish range options:

```js
// Any range option for which _isValidTarget returns false, will be removed 
// from the options object by abstract-leveldown
AbstractLevelDOWN.prototype._isValidTarget = function (target) {
  return target != null
}

EncodingDOWN.prototype._isValidTarget = function (target) {
  return true
}
```

**Comment by vweevers** [comment]
First, find out how and where `ltgt` and also `level-option-wrap` are used. And if there is a situation where `{}` turns into `{ gt: undefined, gte: undefined, .. }`. If not, we're good.

**Comment by vweevers** [comment]
> And if there is a situation where `{}` turns into `{ gt: undefined, gte: undefined, .. }`

Yes. [Run](https://runkit.com/vweevers/5be7fc7ed038f9001282b248):

```js
var wrap = require('level-option-wrap')

// no gt property
var input = {}

var output = wrap(input, {
  gt(x) { return x }
})

if ('gt' in output) {
  console.log('we have a problem')
}
```

Still, I struggle to find a case where this is actually a problem. For example, `subleveldown` uses `level-option-wrap`, but it also [replaces falsy range options with empty strings](https://github.com/Level/subleveldown/blob/f0b5646bde3da53a466cd3da93f2d0526e275e42/leveldown.js#L58), so no problem there.

And the above behavior is arguably a bug in `level-option-wrap`. It should not add properties that don't exist in the input object.

**Comment by vweevers** [comment]
To recap: historically, nullish range options have meant "not defined" in *most* of the ecosystem, but because it's *most* and not *all*, it required a large number of hacks, many undocumented and without integration tests, making it hard to predict behavior, let alone how a change to that behavior will ripple through the ecosystem.

In `abstract-leveldown` v6, nullish range options no longer mean "not defined". Thanks to this, implementations like `encoding-down` can give their own meaning to nullish range options. I'm positive we made the right call there. Though it is scary to change a common behavior in the ecosystem, or rather, to demand a change, I think it would be a mistake to reintroduce "nullish means not defined" in `leveldown` and (by inspiration) every other implementation. The added code complexity counts as yet another hack in my book.

Yes, there is risk of ecosystem breakage, but we have to start somewhere, with a documented, simplest possible behavior, which we did for `abstract-leveldown` and now `leveldown`.

:shipit:
